modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/stat/IgniteStatisticsRepositoryImpl.java

```diff
@@ -45,204 +50,214 @@
      * Constructor.
      *
      * @param storeData If {@code true} - node stores data locally, {@code false} - otherwise.
-     * @param persistence If {@code true} - node have persistence store, {@code false} - otherwise.
-     * @param statisticsManager Ignite statistics manager.
-     * @param log Ignite logger to use.
+     * @param db Database to use in storage if persistence enabled.
+     * @param subscriptionProcessor Subscription processor.
+     * @param statisticsMgr Ignite statistics manager.
+     * @param logSupplier Ignite logger supplier to get logger from.
      */
     public IgniteStatisticsRepositoryImpl(
             boolean storeData,
-            boolean persistence,
-            IgniteStatisticsManagerImpl statisticsManager,
-            IgniteLogger log) {
+            IgniteCacheDatabaseSharedManager db,
+            GridInternalSubscriptionProcessor subscriptionProcessor,
+            IgniteStatisticsManagerImpl statisticsMgr,
+            Function<Class<?>, IgniteLogger> logSupplier
+    ) {
         if (storeData) {
             // Persistence store
-            partsStats = (persistence) ? null : new ConcurrentHashMap<>();
-            localStats = new ConcurrentHashMap<>();
+            store = (db == null) ? new IgniteStatisticsInMemoryStoreImpl(logSupplier) :
+                    new IgniteStatisticsPersistenceStoreImpl(subscriptionProcessor, db, this, logSupplier);
+
+            locStats = new ConcurrentHashMap<>();
         }
         else {
             // Cache only global statistics, no store
-            partsStats = null;
-            localStats = null;
-        }
-        this.statisticsManager = statisticsManager;
-        this.log = log;
-    }
-
-    /**
-     * Convert collection of partition level statistics into map(partId->partStatistics).
-     *
-     * @param key Object key.
-     * @param statistics Collection of tables partition statistics.
-     * @return Partition id to statistics map.
-     */
-    private Map<Integer, ObjectPartitionStatisticsImpl> buildStatisticsMap(
-            StatsKey key,
-            Collection<ObjectPartitionStatisticsImpl> statistics
-    ) {
-        Map<Integer, ObjectPartitionStatisticsImpl> statisticsMap = new ConcurrentHashMap<>();
-        for (ObjectPartitionStatisticsImpl s : statistics) {
-            if (statisticsMap.put(s.partId(), s) != null)
-                log.warning(String.format("Trying to save more than one %s.%s partition statistics for partition %d",
-                        key.schema(), key.obj(), s.partId()));
+            store = null;
```

Comment:
Does it make sense to have dummy repository implementation or may be do not create it on client node at all?

Reply:
"if (store==null) return" in every method look ugly.

Reply:
Yes, we can have a dummy implementation for the client nodes. But there are locStats map too, so part of such ugly tests staing even after that.